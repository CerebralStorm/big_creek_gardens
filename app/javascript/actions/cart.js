import Cookies from 'js-cookie'
import constants from '../constants'
ENV.Cookies = Cookies

let cartKey = function (currentUser = null) {
  if(!!currentUser) {
    return `${currentUser.id}_cart`;
  } else {
    return 'guest_cart';
  }
}

let verifyCart = function (currentUser = null, cart = null) {
  if(typeof(cart) !== 'object') {
    Cookies.set(cartKey(currentUser), {});
    cart = {}
  }
  return cart;
}

let loadCart = function (currentUser = null) {
  return function (dispatch) {
    let cart = Cookies.getJSON(cartKey(currentUser));
    cart = verifyCart(currentUser, cart)
    return dispatch({ type: constants.LOAD_CART, cart: cart })
  }
}

let updateCart = function (currentUser = null, cart = {}) {
  cart = verifyCart(currentUser, cart)
  Cookies.set(cartKey(currentUser), cart);
}

export { loadCart, updateCart };