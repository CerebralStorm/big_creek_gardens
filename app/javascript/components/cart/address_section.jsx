import React from 'react';
import {CardElement} from 'react-stripe-elements';
import SelectUSState from 'react-select-us-states';

class AddressSection extends React.Component {
  render() {
    return (
      <div>
        <div className="form-group m-2">
          <label htmlFor="checkoutCity">Address</label>
          <input type="text" id='checkoutCity' className="form-control" />
        </div>
        <div className='row'>
          <div className="form-group m-2 col-sm-12 col-md-5">
            <label htmlFor="checkoutState">State</label>
            <SelectUSState id="checkoutState" className="form-control" />
          </div>
          <div className="form-group m-2 col-sm-12 col-md-5">
            <label htmlFor="checkoutZip">Zip</label>
            <input type="text" id='checkoutZip' className="form-control" />
          </div>
        </div>
      </div>
    );
  }
}

export default AddressSection;