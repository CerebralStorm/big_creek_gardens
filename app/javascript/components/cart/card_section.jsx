import React from 'react';
import {CardElement} from 'react-stripe-elements';

class CardSection extends React.Component {
  render() {
    return (
      <div className='col-12'>
        <div className="form-group m-2">
          <label htmlFor="checkoutCreditCard" className="col-sm-2 col-form-label">Credit Card</label>
          <CardElement id='checkoutCreditCard' className="form-control" />
        </div>
      </div>
    );
  }
}

export default CardSection;