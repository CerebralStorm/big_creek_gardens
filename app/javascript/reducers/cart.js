import constants from '../constants'

var cartReducer = function (state = {}, action) {
  switch (action.type) {
    case constants.LOAD_CART:
      console.log(state)
      console.log(action.cart)
      return action.cart;
    default:
      return state;
  }
}

export { cartReducer };