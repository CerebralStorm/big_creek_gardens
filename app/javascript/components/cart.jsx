import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { loadCart } from '../actions/cart'
import { Link } from 'react-router-dom'

import CartItem from './cart/cart_item'
import CheckoutForm from './cart/checkout_form'
import { Elements, StripeProvider } from 'react-stripe-elements';

class Cart extends React.Component {
  componentDidMount() {
    this.props.dispatch(loadCart())
  }

  render() {
    return (
      <div className="container cart-container">
        <div className="card shopping-cart">
          <div className="card-header bg-dark text-light">
            <i className="fa fa-shopping-cart" aria-hidden="true"></i>
            Shipping cart
            <Link to={'/shop'} className="btn btn-outline-success btn-sm pull-right">
              Continue Shopping
            </Link>
            <div className="clearfix"></div>
          </div>
          <div class="card-body">
            {Object.values(this.props.cart).map(cartItem => (
              <div>
                <CartItem key={cartItem.product.id} cartItem={cartItem} />
                <hr />
              </div>
            ))}
          </div>
        </div>
        <div class="card-footer">
          <div class="coupon col-md-5 col-sm-5 no-padding-left pull-left">
            <div class="row">
              <div class="col-6">
                <input type="text" class="form-control" placeholder="coupon code" />
              </div>
              <div class="col-6">
                <input type="submit" class="btn btn-default" value="Use coupon" />
              </div>
            </div>
          </div>
          <div class="pull-right" style={{margin: '10px'}}>
            <a href="" class="btn btn-success pull-right">Checkout</a>
            <div class="pull-right total-price" style={{margin: '5px'}}>
              <strong>Total price: <b>$50.00</b></strong>
            </div>
          </div>
          <div className='clearfix'></div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    currentUser: state.currentUser,
    cart: state.cart
  }
}

export default connect(mapStateToProps)(Cart)