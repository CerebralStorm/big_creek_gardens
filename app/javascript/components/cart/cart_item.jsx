import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { updateCart } from '../../actions/cart'

class CartItem extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div className="row">
        <div className="col-12 col-sm-12 col-md-2 text-center">
          <img className="img-responsive" src="http://placehold.it/120x80" alt="prewiew" width="120" height="80" />
        </div>
        <div className="col-12 text-sm-center col-sm-12 text-md-left col-md-6">
          <h4 className="product-name"><strong>{this.props.cartItem.product.name}</strong></h4>
          <h4><small>{this.props.cartItem.product.description}</small></h4>
        </div>
        <div className="col-12 col-sm-12 text-sm-center col-md-4 text-md-right row">
          <div className="col-3 col-sm-3 col-md-6 text-md-right line-item-price">
            <h6><strong>{this.props.cartItem.product.price} <span className="text-muted">x</span></strong></h6>
          </div>
          <div className="col-4 col-sm-4 col-md-4">
            <div className="quantity">
              <input type="button" value="+" className="plus" />
              <input type="number" step="1" max="99" min="1" value="1" title="Qty" className="qty" size="4" />
              <input type="button" value="-" className="minus" />
            </div>
          </div>
          <div className="col-2 col-sm-2 col-md-2 text-right">
            <button type="button" className="btn btn-outline-danger btn-xs">
              <i className="fa fa-trash" aria-hidden="true"></i>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    currentUser: state.currentUser,
    cart: state.cart
  }
}

export default connect(mapStateToProps)(CartItem)

