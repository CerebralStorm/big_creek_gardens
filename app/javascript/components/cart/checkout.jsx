import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { updateCart, loadCart } from '../../actions/cart'

import CheckoutForm from './checkout_form'
import { Elements, StripeProvider } from 'react-stripe-elements';

class Checkout extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <StripeProvider apiKey={ENV.stripe_key} >
        <Elements>
          <CheckoutForm />
        </Elements>
      </StripeProvider>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    currentUser: state.currentUser,
    cart: state.cart
  }
}

export default connect(mapStateToProps)(Checkout)

