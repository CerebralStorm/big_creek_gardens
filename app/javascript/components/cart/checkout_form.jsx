import React from 'react';
import {CardElement, injectStripe} from 'react-stripe-elements';
import { updateCart, loadCart } from '../../actions/cart'
import ChargeApi from '../../api/charge_api'
import { Redirect } from 'react-router';
import UserSection from './user_section';
import AddressSection from './address_section';
import CardSection from './card_section';

class CheckoutForm extends React.Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.totalPrice = this.totalPrice.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleStateChange = this.handleStateChange.bind(this);
    this.state = {
      name: '',
      email: '',
      phone: '',
      address: '',
      city: '',
      state: '',
      zip: '',
      redirectToConfirmation: false,
      orderId: ''
    }
  }

  totalPrice() {
    let total = 0;
    Object.values(this.props.cart).forEach((item, index) => {
      total += (item.quantity * item.product.price)
    })
    return total.toFixed(2)
  }

  handleChange(event) {
    let change = {}
    change[event.target.name] = event.target.value
    this.setState(change)
  }

  handleStateChange(state) {
    this.setState({state: state})
  }

  async handleSubmit(event) {
    event.preventDefault()
    let user = {
      name: this.state.name,
      email: this.state.email,
      phone: this.state.phone,
      address_line1: this.state.address,
      address_city: this.state.city,
      address_state: this.state.state,
      address_zip: this.state.zip,
      address_country: 'US',
    }

    let orderlineItems = Object.values(this.props.cart).map(cartItem => {
      return ({
        product_id: cartItem.product.id,
        quantity: cartItem.quantity
      });
    })

    let {token} = await this.props.stripe.createToken(user);
    ChargeApi.createCharge({
      stripe_token: token.id,
      user: user,
      order_line_items_attributes: orderlineItems,
      authenticity_token: ENV.csrf_token
    }).then((response) => {
      updateCart({})
      console.log(response.data)
      this.setState({orderId: response.data.response.order_id, redirectToConfirmation: true})
    })
  }

  render() {
    if (this.state.redirectToConfirmation) {
      return <Redirect to={`/confirmation/${this.state.orderId}`} />;
    }

    return (
      <div className='container checkout-container'>
        <h1 className='text-center'>Checkout</h1>
        <form onSubmit={this.handleSubmit} className='m-4'>
          <div className='row'>
            <div className='col-sm-12 col-md-8 card pb-3'>
              <UserSection handleChange={this.handleChange} name={this.state.name} email={this.state.email} phone={this.state.phone} />
              <AddressSection handleStateChange={this.handleStateChange} handleChange={this.handleChange} address={this.state.address} state={this.state.state} zip={this.state.zip} />
              <CardSection />
              <button onClick={this.handleSubmit} className='btn btn-success pull-right'>Confirm order</button>
            </div>
            <div className='col-sm-12 col-md-4'>
              <ul className='list-group mb-3'>
                <li className="list-group-item checkout-cart-header"><h6 className='text-center checkout-title'>Cart</h6></li>
              {Object.values(this.props.cart).map(cartItem => (
                <li key={cartItem.product.id} className="list-group-item d-flex justify-content-between lh-condensed">
                  <div className='col-9 checkout-item-small'>
                    <h6 className="checkout-name">{cartItem.product.name}</h6>
                    <small className="text-muted checkout-description">{cartItem.product.description}</small>
                  </div>
                  <span className="text-muted">${(cartItem.product.price * cartItem.quantity).toFixed(2)}</span>
                </li>
              ))}
              <li className='list-group-item d-flex justify-content-between lh-condensed checkout-total'>
                <span>
                  <h6>Total</h6>
                </span>
                <h6 className='text-center'><strong>${this.totalPrice()}</strong></h6>
              </li>
              </ul>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

export default injectStripe(CheckoutForm);