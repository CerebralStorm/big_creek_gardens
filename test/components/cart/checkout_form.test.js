import React from 'react'
import { CheckoutForm } from '../../../app/javascript/components/cart/checkout_form'

let match = {
  params: {
    orderId: 1
  }
}

let currentUser = {
  id: 1,
  name: 'Steve'
}

let cart = {};

test('render a CheckoutForm', () => {
  const wrapper = shallow(<CheckoutForm match={match} currentUser={currentUser} cart={cart} />);
  expect(wrapper).toMatchSnapshot();
});