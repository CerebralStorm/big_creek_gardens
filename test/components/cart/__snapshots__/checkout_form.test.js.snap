// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a CheckoutForm 1`] = `
<div
  className="container checkout-container"
>
  <h1
    className="text-center"
  >
    Checkout
  </h1>
  <form
    className="m-4"
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12 col-md-8 card pb-3"
      >
        <UserSection
          handleChange={[Function]}
          name="Steve"
        />
        <AddressSection
          handleChange={[Function]}
          handleStateChange={[Function]}
        />
        <CardSection />
        <button
          className="btn btn-success pull-right"
          disabled={true}
          onClick={[Function]}
        >
          Confirm order
        </button>
      </div>
      <div
        className="col-sm-12 col-md-4"
      >
        <ul
          className="list-group mb-3"
        >
          <li
            className="list-group-item checkout-cart-header"
          >
            <h6
              className="text-center checkout-title"
            >
              Cart
            </h6>
          </li>
          <li
            className="list-group-item d-flex justify-content-between lh-condensed checkout-total"
          >
            <span>
              <h6>
                Total
              </h6>
            </span>
            <h6
              className="text-center"
            >
              <strong>
                $
                0.00
              </strong>
            </h6>
          </li>
        </ul>
      </div>
    </div>
  </form>
</div>
`;
